### ISSUE TITLE
Large memory spikes when calling "HeapProfiler.takeSnapshot"

### ISSUE THREAD

**Issue by paulrutter** [ChromeDevTools/devtools-protocol#71]
# Problem

Recently, i created a Node.js module which uses the DevTools protocol to create heapdumps of a Node.js process. This works fine except that calling "HeapProfiler.takeSnapshot" generates a very large memory spike.

# Analysis
I'm not sure why this happens, but my guess is that the whole memory snapshot is sent over the socket as fast as possible, but the client (my Node.js app) cannot handle the data that fast, causing in-memory buffering. I know WebSockets do not have a backpressure mechanism, so this may be hard to solve. An alternative could be to stream the heapsnapshot over the HTTP protocol, so the default Node.js streams can handle backpressure.

# Question
Long story short: are there any known issues with "[HeapProfiler.takeSnapshot](https://chromedevtools.github.io/devtools-protocol/v8/HeapProfiler/#method-takeHeapSnapshot)" in relation to memory consumption? If so, how should i handle this?

# Reproduction path / code
Code can be viewed [here](https://github.com/blueconic/node-oom-heapdump/blob/master/lib/heapdumpWorker.js).
It can be reproduced by calling `npm test` in the root of the project, once git cloned. 
Once you see "Started heapdump OoM worker ..." in the console, take a look at the memory footprint in `top` on Linux, or `process manager` on Windows. It peaks at over 2GB.

Thanks in advance!

**Comment by a1ph** [comment]
Hi Paul. This is a known problem with heap snapshots. Last time I checked it was not about websockets. The data is buffered in the DevTools front-end between main thread and heap snapshot worker that parses it.

**Comment by paulrutter** [comment]
Ah right, thanks for your swift response. Any workaround besides increasing memory? I'm running into OoM killer in Docker, because container memory is exceeded.

Do you know if there is an issue i can track progress in? Thanks!

**Comment by a1ph** [comment]
Sorry, no workaround besides increasing the heap limit. :-( You can watch crbug.com/768355 for the progress.

**Comment by paulrutter** [comment]
Ok lets continue discussion in that issue. Thanks!
